### DISCUSSION TITLE
Possible improvements

### DISCUSSION THREAD

**Discussion by Darel13712** [sberbank-ai-lab/RePlay#47] [Ideas]
# Cold Scenario
Having a cold model inside **every** scenario is clunky and unneccessary, this should be extracted into a separate class similar to `Fallback`.

# Two Stages
The code scares me, it should be refactored into separate logical functions. I think one should be able to train second level reranking, having trained base models. (i.e. without providing splitters and training everything from scratch). This will also allow for separate retraining of first level and second level.

# Data preparator
It currently provides questionable utility and can be safely removed. It is not clear why do we need it atm.

# Indexer
There are a lot of conversions of indexers in every model. One way to solve this is to force user to convert indexes before training the models and convert back after getting predictions. We will have to store extra object, but model logic will be eased.

# Ignite
We don't actually profit much from having ignite for neural models and we can probably benefit if we write our own training loop. This will probably make the code clearer.

**Comment by monkey0head** [comment]
**Two-Stages**
- agree, pipeline could be divided into customisable parts. we should discuss the parts and their interface
- we can try to use catboost for spark for now as a second level is it is spark and it does not do any unclear data preprocessing

**DataPreparator**
- we could leave a function to convert column to timestamp
- we could remove datapreparator, but add a function to check if log/features dataframe is valid (no nulls in ids, has necessary columns, no duplicates / warn on duplicates)

**Indexers**
Some models require indexers (als/slim/multvae/lightfm), оther dо not. We can try add an attribute, indicating if reindexing is required.
We can also force the user to convert indexes but we will need to check cold items/users appearance

**DSSM/neural architecture with features**
We just need it 

**Pandas as an input**
No pandas, pls. Providing data in pandas and conversion with convert2spark leads to implicit types conversion. e.g. pandas reads ints as float if there are nulls in a column and spark does not. There are also some problems with `long` ids if we do not define type explicitly in schema. I am not able to describe clearly (
